### PR TITLE
feature: nixos module reuse to build arbitrary hermes module for chain pair (not only on/for nixos/serviced)

### PR DESCRIPTION
**Pull request by dzmitry-lahoda**

### CODE REVIEW

**Comment by JonathanLorimer** [comment]
Just wanted to comment that I am really happy to see someone (who probably has way more knowledge about cosmos chains and relayers than me) working on this. My dream for this project has always been to improve cosmos infrastructure by bringing the determinism and declarative-ness of nix.

If you want to pair, or want any help with the nix side of things (probably the only place I can really help), just let me know. 

Also if you want help testing this with NixOS tests I can help there too.

Thank you!

**Comment by dzmitry-lahoda** [comment]
hello, sure :)


1. i split options to files.
2. old options named old. 
3. so no old users are broken.
4. i made new generator
5. its generator is 1 to 1 with TOML of Hermes (since 1.6)
6. make same split.
7. so that it can be used in any module (i do not use nixos/serviced, but process compose).

**Comment by dzmitry-lahoda** [comment]
removed old config

**Comment by dzmitry-lahoda** [comment]
@JonathanLorimer seems i am done. so not sure I have run qemu tests.

File: examples/hermes-config.nix [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                    denominator = "3";
+                  };
+                }
+              ];
+            }
+          ];
+        })
+        .config
+        .hermes
+        .toml;
```
Comment (JonathanLorimer) [inline examples/hermes-config.nix:42]: This is very cool

File: nixosModules/relayer/hermes.nix [hunk 2]
```diff
@@ -2,407 +2,30 @@
...
-            port = mkOption {
-              type = types.port;
-              default = 3001;
-              description = ''
-                Specify the port over which the built-in HTTP server will serve the metrics gathered.
-              '';
-            };
-          };
+    options.services.hermes =
+      import ./options.nix {inherit lib;}
```
Comment (JonathanLorimer) [inline nixosModules/relayer/hermes.nix:17]: This is really clean, nice.

**Review by JonathanLorimer — COMMENTED** [review]
This looks great, just a couple of comments

File: flake.nix [hunk 3]
```diff
@@ -26,6 +26,7 @@
         ./modules/apps.nix
         # Sets the overlays.default flake output
         ./modules/overlay.nix
+        ./examples/hermes-config.nix
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply flake.nix:29]: sure, package is bad.

are you aware of any convetion here? like how to run nixosMules? is name standard?

how to run generator tests in this repo? I use example as test btw.

[the remark this answers is not in the document]
Comment (JonathanLorimer) [reply flake.nix:29]: Here is some documentation from flake parts https://flake.parts/options/flake-parts.html?highlight=nixosModules#opt-flake.nixosModules 

Its a standard output on flakes https://nixos.wiki/wiki/Flakes#Output_schema. It allows you to export nixosModules for consumption.

**Comment by dzmitry-lahoda** [comment]
- seems can improve some type error messages https://github.com/NixOS/nixpkgs/blob/master/lib/types.nix  . at least either for event_source, some number = int | float, and may be allow numbers to be coersed to sting when needed, and must have value, non empty single line sting (seem all types are existing)

**Comment by dzmitry-lahoda** [comment]
seems nice approach to give users toml files from modules - see yaml from k8s https://github.com/hall/kubenix . will refactor to that

File: nixosModules/relayer/hermes.nix [hunk 4]
```diff
@@ -2,407 +2,30 @@
   lib,
   config,
   pkgs,
+  nix-std ? import (fetchTarball "https://github.com/chessai/nix-std/archive/master.tar.gz"),
   ...
 }: let
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply nixosModules/relayer/hermes.nix:7]: ok will do

**Comment by dzmitry-lahoda** [comment]
working on making this done

**Comment by JonathanLorimer** [comment]
@dzmitry-lahoda just ping me when you want a review! Also feel free to ask for a review before you done if you want eyes on anything.

**Comment by dzmitry-lahoda** [comment]
closing this one, will open new PR with fixed comments (and more tests).

File: lib/default.nix [hunk 5]
```diff
@@ -140,4 +140,31 @@ in {
       install_name_tool -add_rpath "${libwasmvm}/lib" $out/bin/${binName}
     ''
     else null;
+
+  evalHermesModule = {modules}:
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply lib/default.nix:168]: https://github.com/terranix/terranix/blob/fc9077ca02ab5681935dbf0ecd725c4d889b9275/core/default.nix#L49

> hermesModuleConfigToml

will use this

**Pull request closed without merging**